### PR TITLE
Bandwidth test pr

### PR DESCRIPTION
**Pull request by wangbluo**
Main feature and the fix for all_gather, broadcast, all_to_all collective op bandwidth computation. 
img_v3_027b_aa0f9c52-ce1a-40b0-a9c3-e0096d478e7g
img_v3_027b_c147157f-6fa3-45bd-9f8d-1f23c6f5fe7g
img_v3_027b_236ccabe-af10-4343-96f0-feeb153c08ag

multi nodes multi gpus：
img_v3_027b_44e5f429-6e6f-4912-8151-aaa6825030cg
img_v3_027b_484d8993-21bc-4671-b624-2a8b6e1cf58g

### CODE REVIEW

File: test_bandwidth.py [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+os.environ['CUDA_VISIBLE_DEVICES'] = '1,2,3,4'
+    
+def setup(rank, world_size):
+    os.environ["MASTER_ADDR"] = os.getenv("MASTER_ADDR", "localhost")
+    os.environ["MASTER_PORT"] = os.getenv("MASTER_PORT", "12355")
+    os.environ["RANK"] = os.getenv("RANK", "0")
+    os.environ["WORLD_SIZE"] = os.getenv("WORLD_SIZE", "1")
+    dist.init_process_group(world_size=world_size, rank=rank,
+                            init_method="env://", backend="nccl")
+    torch.cuda.set_device(rank)
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: This is not right, what if the rank is more than 8, i.e. I am running with 16 GPUs.

File: test_bandwidth.py [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+import sys
+import time
+import argparse
+from abc import abstractmethod
+
+import torch
+import torch.distributed as dist
+from typing import List
+from prettytable import PrettyTable
+os.environ['CUDA_VISIBLE_DEVICES'] = '1,2,3,4'
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: You should not set this in the script. You should just do it in shell.

File: test_bandwidth.py [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+class AllToAll(CollectiveOps):
+    def __call__(self, tensor: torch.Tensor) -> None:
+        input_tensor_list = list(torch.chunk(tensor, self.world_size, dim=0))
+        output_tensor_list = input_tensor_list[:]
+        dist.all_to_all(output_tensor_list, input_tensor_list)
+
+    def ranks_factor(self):
+        return (self.world_size - 1) / self.world_size 
+
+def avg_time(op, tensor:torch.tensor, iters):
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: `torch.tensor` is a function, not a class.

Comment (FrankLeeeee) [reply test_bandwidth.py]: If you have type hinting for one arg, you should add hinting for other args as well.

File: test_bandwidth.py (hunk 1 above)
Comment (wangbluo) [reply test_bandwidth.py]: Okay, got it, I'll fix it.

File: test_bandwidth.py (hunk 2 above)
Comment (wangbluo) [reply test_bandwidth.py]: I see, I'll move it to the shell.

File: test_bandwidth.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+    def __call__(self, tensor: torch.Tensor) -> None:
+        pass
+
+class AllReduce(CollectiveOps):
+    def __call__(self, tensor: torch.Tensor) -> None:
+        tensor = tensor.contiguous()
+        dist.all_reduce(tensor)
+    
+    def ranks_factor(self):
+        return 2 * (self.world_size - 1) / self.world_size
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: if every subclass of `CollectiveOps` implements `ranks_factor`, shouldn't it be an abstract method in `CollectiveOps`? Meanwhile, what does `ranks_factor` mean?

File: test_bandwidth.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+    Broadcast:
+        B = S/t = algbw 
+    AllToAll:
+        B = S/t * (n-1)/n = algbw * (n-1)/n
+"""
+
+class CollectiveOps:
+    @abstractmethod
+    def __init__(self, world_size: int):
+        self.world_size = world_size
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: Why would this be an abstract method?

File: test_bandwidth.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    AllGather:
+        B = S/t * (n-1)/n = algbw * (n-1)/n
+    Broadcast:
+        B = S/t = algbw 
+    AllToAll:
+        B = S/t * (n-1)/n = algbw * (n-1)/n
+"""
+
+class CollectiveOps:
+    @abstractmethod
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: Just to be precise, if you use `Ops`, it means many operations. I guess you want to use `Op` instaed.

File: test_bandwidth.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        bytes = bytes[:-1]
+    elif bytes[-1] == 'K':
+        return int(bytes[:-1]) * 1024
+    elif bytes[-1] == 'M':
+        return int(bytes[:-1]) * 1024**2
+    elif bytes[-1] == 'G':
+        return int(bytes[:-1]) * 1024**3
+    return int(bytes)
+
+def get_tensor_list(minbytes:str, maxbytes:str, stepbytes:str, stepfactor:str, dtype = torch.float) -> List[int]:
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: user snake case for readability, e.g. stepfactor -> step_factor.

File: test_bandwidth.py (hunk 4 above)
Comment (wangbluo) [reply test_bandwidth.py]: Yes, sure. I forgot it. 
In the above comments regarding bandwidth computation, 'n' represents the number of ranks, and all factors are totally about 'n'."

File: test_bandwidth.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    parser.add_argument('-d', '--dtype', type=str,default='float', 
+                        choices=['fp16', 'bf16', 'float'])
+    args = parser.parse_args()
+
+    #setup
+    if torch.cuda.is_available():
+        print("CUDA is available. GPU can be used.")
+    else:
+        print("CUDA is not available. Using CPU.")
+        sys.exit("Exiting the program due to lack of GPU.")
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: This looks weird to me, if you are telling the user that CPU will be used, why does this program exit?

File: test_bandwidth.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+from typing import List
+from prettytable import PrettyTable
+os.environ['CUDA_VISIBLE_DEVICES'] = '1,2,3,4'
+    
+def setup(rank, world_size):
+    os.environ["MASTER_ADDR"] = os.getenv("MASTER_ADDR", "localhost")
+    os.environ["MASTER_PORT"] = os.getenv("MASTER_PORT", "12355")
+    os.environ["RANK"] = os.getenv("RANK", "0")
+    os.environ["WORLD_SIZE"] = os.getenv("WORLD_SIZE", "1")
+    dist.init_process_group(world_size=world_size, rank=rank,
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: You have two sources for rank and world_size:
1. from the function argument
2. from environment

what's more is that the function argument is also fetched from environment variable, this makes this code snippet rather messy and error-prone.

File: test_bandwidth.py (hunk 7 above)
Comment (wangbluo) [reply test_bandwidth.py]: image
https://github.com/NVIDIA/nccl-tests
I saw the stepbytes and stepfactor parameters in the nccl tests's read.me, so I just want to keep it same. I'll fix it.

File: run_test_bandwidth.sh [hunk 10]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+#sh run_test_bandwidth.sh 2 2 0 10.20.1.81 22 
+set -e
```
Comment (FrankLeeeee) [inline run_test_bandwidth.sh]: Can you provide a command example for launching this script for single-node multi-gpu and multi-node multi-GPU testing in README?

Comment (wangbluo) [reply run_test_bandwidth.sh]: Okay, sure. 
Just use the command in the second node GPU, change the $NODE_RANK parameter to 1, like:
sh run_test_bandwidth.sh 2 2 1 10.20.1.81 22 
And $NNODES represents the number of nodes.

File: test_bandwidth.py (hunk 3 above)
Comment (wangbluo) [reply test_bandwidth.py]: Ok, I'll remove it

File: test_bandwidth.py (hunk 5 above)
Comment (wangbluo) [reply test_bandwidth.py]: Sorry, it shouldn't. I remove it.

File: test_bandwidth.py (hunk 8 above)
Comment (wangbluo) [reply test_bandwidth.py]: Okay, I'll remove it. So it will be the bandwidth of CPUs.

File: test_bandwidth.py (hunk 9 above)
Comment (wangbluo) [reply test_bandwidth.py]: Sorry, I'll remove the args of the function argument.

File: run_test_bandwidth.sh [hunk 11]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+#sh run_test_bandwidth.sh 2 2 0 10.20.1.81 22
+#sh run_test_bandwidth.sh 2 2 1 10.20.1.81 22
```
Comment (FrankLeeeee) [inline run_test_bandwidth.sh]: Encountered this problem while executing the command.

Screenshot 2024-01-23 at 14 08 20

File: test_bandwidth.py [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+    setup()
+
+    if args.dtype == 'fp16':
+        dtype = torch.float16
+    elif args.dtype == 'bf16':
+        dtype = torch.bfloat16
+    elif args.dtype == 'float':
+        dtype = torch.float
+        
+    world_size = int(os.getenv("WORLD_SIZE", 1))
```
Comment (FrankLeeeee) [inline test_bandwidth.py]: use `dist.get_world_size`.

**Review by FrankLeeeee — APPROVED** [review]

**Pull request merged**